### ISSUE TITLE
Feature: Specify Exact Refresh Rate

### ISSUE THREAD

**Issue by GrabbenD** [alex-courtis/way-displays#128]
# Issue
After starting the compositor the screen goes black and half of it is visible if I move my mouse. If I start a application like foot terminal emulator the screen blinks black when nothing is moving. As soon as I launch a game everything works as intended.

It seems like damage tracking is acting up?

# Setup
- Version (Way-Displays): 1.8.2
- Compositor: river-git
- Distro: Arch Linux
- GPU: RX 6800 XT

# River config (`~/.config/river/init`)
Default (https://github.com/riverwm/river/blob/master/example/init) except for:
```bash
way-displays > /tmp/way-displays.${XDG_VTNR}.${USER}.log 2>&1 &
```

# Way-Displays config (`~/.config/way-displays/cfg.yaml`)
Default except for:
```
MODE:
    - NAME_DESC: DP-2
      WIDTH: 3840
      HEIGHT: 1600
      HZ: 119.982
```

**Comment by alex-courtis** [comment]
That's no good... everything should work on river. I see you have also opened #129 

I'll ask some questions in both issues, which should apply to both.

**Comment by alex-courtis** [comment]
What happens when you run the [sway](https://github.com/swaywm/sway) compositor? You can add way-displays as per https://github.com/alex-courtis/way-displays#sway

That should give us some insight as to whether this is an amdgpu/wlroots/mesa/vulkan issue vs a river issue.

**Comment by GrabbenD** [comment]
Thanks for taking a look at this!

> I'll ask some questions in both issues, which should apply to both.

Let's start with this issue! **I realized that I get a similar issue when gaming with VRR if my refresh rate is wrong.**

> What happens when you run the [sway](https://github.com/swaywm/sway) compositor? You can add way-displays as per https://github.com/alex-courtis/way-displays#sway

Seems like way-displays enables VRR (as that's the default) as intended but my refresh rate is wrong. I'm running this command after starting Sway with Way-Displays 1.8.2
```console
> wlr-randr
DP-3 "Dell Inc. Dell AW3821DW #HLAYMxgwABTR (DP-3)"
  Make: Dell Inc.
  Model: Dell AW3821DW
  Serial: #HLAYMxgwABTR
  Physical size: 880x370 mm
  Enabled: yes
  Modes:
    3840x1600 px, 59.993999 Hz (preferred, current)
    3840x1600 px, 143.998001 Hz
    3840x1600 px, 120.044998 Hz
    3840x1600 px, 119.982002 Hz
    3840x1600 px, 99.970001 Hz
    3840x1600 px, 96.036003 Hz
    3840x1600 px, 84.971001 Hz
    3840x1600 px, 72.019997 Hz
    3840x1600 px, 60.007999 Hz
    3840x1600 px, 50.002998 Hz
    3840x1600 px, 48.008999 Hz
    3840x1600 px, 29.971001 Hz
    3840x1600 px, 30.000000 Hz
    3840x1600 px, 25.000999 Hz
    3840x1600 px, 23.976000 Hz
    3840x1600 px, 24.002001 Hz
    1920x1200 px, 59.993999 Hz
    1920x1080 px, 59.993999 Hz
    1600x1200 px, 59.993999 Hz
    1680x1050 px, 59.993999 Hz
    1280x1024 px, 59.993999 Hz
    1440x900 px, 59.993999 Hz
    1280x800 px, 59.993999 Hz
    1280x720 px, 59.993999 Hz
    1024x768 px, 60.004002 Hz
    800x600 px, 60.317001 Hz
    640x480 px, 60.000000 Hz
    640x480 px, 59.939999 Hz
  Position: 0,0
  Transform: normal
  Scale: 1.000000
  Adaptive Sync: enabled
```

For reference:
<details>
  <summary>swaymsg -t get_outputs</summary>
  
  ```json
  [
    {
      "id": 3,
      "type": "output",
      "orientation": "none",
      "percent": 1.0,
      "urgent": false,
      "marks": [
      ],
      "layout": "output",
      "border": "none",
      "current_border_width": 0,
      "rect": {
        "x": 0,
        "y": 0,
        "width": 3840,
        "height": 1600
      },
      "deco_rect": {
        "x": 0,
        "y": 0,
        "width": 0,
        "height": 0
      },
      "window_rect": {
        "x": 0,
        "y": 0,
        "width": 0,
        "height": 0
      },
      "geometry": {
        "x": 0,
        "y": 0,
        "width": 0,
        "height": 0
      },
      "name": "DP-3",
      "window": null,
      "nodes": [
      ],
      "floating_nodes": [
      ],
      "focus": [
        4
      ],
      "fullscreen_mode": 0,
      "sticky": false,
      "primary": false,
      "make": "Dell Inc.",
      "model": "Dell AW3821DW",
      "serial": "#HLAYMxgwABTR",
      "modes": [
        {
          "width": 3840,
          "height": 1600,
          "refresh": 59994,
          "picture_aspect_ratio": "none"
        },
        {
          "width": 3840,
          "height": 1600,
          "refresh": 143998,
          "picture_aspect_ratio": "none"
        },
        {
          "width": 3840,
          "height": 1600,
          "refresh": 120045,
          "picture_aspect_ratio": "none"
        },
        {
          "width": 3840,
          "height": 1600,
          "refresh": 119982,
          "picture_aspect_ratio": "none"
        },
        {
          "width": 3840,
          "height": 1600,
          "refresh": 99970,
          "picture_aspect_ratio": "none"
        },
        {
          "width": 3840,
          "height": 1600,
          "refresh": 96036,
          "picture_aspect_ratio": "none"
        },
        {
          "width": 3840,
          "height": 1600,
          "refresh": 84971,
          "picture_aspect_ratio": "none"
        },
        {
          "width": 3840,
          "height": 1600,
          "refresh": 72020,
          "picture_aspect_ratio": "none"
        },
        {
          "width": 3840,
          "height": 1600,
          "refresh": 60008,
          "picture_aspect_ratio": "none"
        },
        {
          "width": 3840,
          "height": 1600,
          "refresh": 50003,
          "picture_aspect_ratio": "none"
        },
        {
          "width": 3840,
          "height": 1600,
          "refresh": 48009,
          "picture_aspect_ratio": "none"
        },
        {
          "width": 3840,
          "height": 1600,
          "refresh": 29971,
          "picture_aspect_ratio": "none"
        },
        {
          "width": 3840,
          "height": 1600,
          "refresh": 30000,
          "picture_aspect_ratio": "none"
        },
        {
          "width": 3840,
          "height": 1600,
          "refresh": 25001,
          "picture_aspect_ratio": "none"
        },
        {
          "width": 3840,
          "height": 1600,
          "refresh": 23976,
          "picture_aspect_ratio": "none"
        },
        {
          "width": 3840,
          "height": 1600,
          "refresh": 24002,
          "picture_aspect_ratio": "none"
        },
        {
          "width": 1920,
          "height": 1200,
          "refresh": 59994,
          "picture_aspect_ratio": "none"
        },
        {
          "width": 1920,
          "height": 1080,
          "refresh": 59994,
          "picture_aspect_ratio": "none"
        },
        {
          "width": 1600,
          "height": 1200,
          "refresh": 59994,
          "picture_aspect_ratio": "none"
        },
        {
          "width": 1680,
          "height": 1050,
          "refresh": 59994,
          "picture_aspect_ratio": "none"
        },
        {
          "width": 1280,
          "height": 1024,
          "refresh": 59994,
          "picture_aspect_ratio": "none"
        },
        {
          "width": 1440,
          "height": 900,
          "refresh": 59994,
          "picture_aspect_ratio": "none"
        },
        {
          "width": 1280,
          "height": 800,
          "refresh": 59994,
          "picture_aspect_ratio": "none"
        },
        {
          "width": 1280,
          "height": 720,
          "refresh": 59994,
          "picture_aspect_ratio": "none"
        },
        {
          "width": 1024,
          "height": 768,
          "refresh": 60004,
          "picture_aspect_ratio": "none"
        },
        {
          "width": 800,
          "height": 600,
          "refresh": 60317,
          "picture_aspect_ratio": "none"
        },
        {
          "width": 640,
          "height": 480,
          "refresh": 60000,
          "picture_aspect_ratio": "4:3"
        },
        {
          "width": 640,
          "height": 480,
          "refresh": 59940,
          "picture_aspect_ratio": "none"
        }
      ],
      "non_desktop": false,
      "active": true,
      "dpms": true,
      "power": true,
      "scale": 1.0,
      "scale_filter": "nearest",
      "transform": "normal",
      "adaptive_sync_status": "enabled",
      "current_workspace": "1",
      "current_mode": {
        "width": 3840,
        "height": 1600,
        "refresh": 59994,
        "picture_aspect_ratio": "none"
      },
      "max_render_time": 0,
      "focused": true,
      "subpixel_hinting": "unknown"
    }
  ]
  ```
  
</details>



Here's my second attempt at setting the right mode:
```
# Default (`~/.config/way-displays/cfg.yaml`) except for:
ORDER:
  - 'DP-3'

MODE:
    # Resolution and refresh
    - NAME_DESC: DP-3
      WIDTH: 3840
      HEIGHT: 1600
      HZ: 119.982002

#VRR_OFF:
#   - DP-3
```

**I can disable VRR if I uncomment the last 2 lines!**
Am I setting the refresh rate wrong?

**Comment by GrabbenD** [comment]
I'm actually able to set the refresh rate if I do `HZ: 120` with both `NAME_DESC: DP-3` and `NAME_DESC: 'Dell Inc. Dell AW3821DW #HLAYMxgwABTR'` 
```
I [11:08:55] DP-3 Changing:
I [11:08:55]   from:
I [11:08:55]     scale:    1.000 (1.149)
I [11:08:55]     position: 0,0
I [11:08:55]     mode:     3840x1600@60Hz (59,994mHz) (preferred)
I [11:08:55]     VRR:      off
I [11:08:55]   to:
I [11:08:55]     mode:     3840x1600@120Hz (120,045mHz)
I [11:08:55]
I [11:08:55] Changes successful
```

Luckily it doesn't use a round number which would otherwise trigger a MCLK bug in AMDGPU (https://gitlab.freedesktop.org/drm/amd/-/issues/2657)

**Nonetheless, is it possible to use use `119,982mHz` instead of `120,045mHz`?**

**Comment by alex-courtis** [comment]
> Nonetheless, is it possible to use use 119,982mHz instead of 120,045mHz?

Unfortunately user defined Hz is an integer and in your case will be rounded to 120. Tt will use the higher value 120,045 and would only fall back to 119,982 on failure.

This can be changed: monitors specify refresh as an even number of milliHz. The pragmatic means to achieve this would simply be to change HZ in cfg.yaml and CLI to be a decimal rounded to 3 D.P.

**Comment by alex-courtis** [comment]
Are you able to use 144Hz? Monitors are not always happy with some modes, despite their advertising them.

**Comment by alex-courtis** [comment]
wlr-randr appears to be abandoned; emersion replaced it with [kanshi](https://sr.ht/~emersion/kanshi/) as the randr model doesn't really work well in the wayland world.

**Comment by GrabbenD** [comment]
Due to bandwidth limitation of DP 1.4 and lack of DSC in my monitor I can either choose between:
- 8 bit 144HZ
- 10 bit 120 Hz

For mysterious reasons I sometimes get unstable screen brightness (flicker) with 120.045 mHz, I found 119.982 mHz to work the best!

The ability to switch to precise modes like with Sway/Hyprland would be highly appreciated

**Comment by alex-courtis** [comment]
@GrabbenD  I'd be most grateful if you could test this feature.

Your exact refresh in cfg.yaml `HZ: 119.982` wil be applied if available.

Also available via CLI `way-displays --set MODE 'Dell AW3821DW' 3840 1600 119.982`

```sh
git clone git@github.com:alex-courtis/way-displays.git
cd way-displays
git checkout 128-exact-refresh
git submodule update --init
make
sudo make install
```

When you are done you can
```
sudo make uninstall
```

**Comment by GrabbenD** [comment]
Great work @alex-courtis 
Works just as intended:

`HZ: 144` (works)

<details>
  <summary>wlr-randr</summary>
  
  ```
  DP-3 "Dell Inc. Dell AW3821DW #HLAYMxgwABTR (DP-3)"
    Make: Dell Inc.
    Model: Dell AW3821DW
    Serial: #HLAYMxgwABTR
    Physical size: 880x370 mm
    Enabled: yes
    Modes:
      3840x1600 px, 59.993999 Hz (preferred)
      3840x1600 px, 143.998001 Hz (current)
      3840x1600 px, 120.044998 Hz
      3840x1600 px, 119.982002 Hz
      3840x1600 px, 99.970001 Hz
      3840x1600 px, 96.036003 Hz
      3840x1600 px, 84.971001 Hz
      3840x1600 px, 72.019997 Hz
      3840x1600 px, 60.007999 Hz
      3840x1600 px, 50.002998 Hz
      3840x1600 px, 48.008999 Hz
      3840x1600 px, 29.971001 Hz
      3840x1600 px, 30.000000 Hz
      3840x1600 px, 25.000999 Hz
      3840x1600 px, 23.976000 Hz
      3840x1600 px, 24.002001 Hz
      1920x1200 px, 59.993999 Hz
      1920x1080 px, 59.993999 Hz
      1600x1200 px, 59.993999 Hz
      1680x1050 px, 59.993999 Hz
      1280x1024 px, 59.993999 Hz
      1440x900 px, 59.993999 Hz
      1280x800 px, 59.993999 Hz
      1280x720 px, 59.993999 Hz
      1024x768 px, 60.004002 Hz
      800x600 px, 60.317001 Hz
      640x480 px, 60.000000 Hz
      640x480 px, 59.939999 Hz
    Position: 0,0
    Transform: normal
    Scale: 1.000000
    Adaptive Sync: enabled
  ```
  
</details>

`HZ: 143.998001` (works)

<details>
  <summary>wlr-randr</summary>
  
  ```
  DP-3 "Dell Inc. Dell AW3821DW #HLAYMxgwABTR (DP-3)"
    Make: Dell Inc.
    Model: Dell AW3821DW
    Serial: #HLAYMxgwABTR
    Physical size: 880x370 mm
    Enabled: yes
    Modes:
      3840x1600 px, 59.993999 Hz (preferred)
      3840x1600 px, 143.998001 Hz (current)
      3840x1600 px, 120.044998 Hz
      3840x1600 px, 119.982002 Hz
      3840x1600 px, 99.970001 Hz
      3840x1600 px, 96.036003 Hz
      3840x1600 px, 84.971001 Hz
      3840x1600 px, 72.019997 Hz
      3840x1600 px, 60.007999 Hz
      3840x1600 px, 50.002998 Hz
      3840x1600 px, 48.008999 Hz
      3840x1600 px, 29.971001 Hz
      3840x1600 px, 30.000000 Hz
      3840x1600 px, 25.000999 Hz
      3840x1600 px, 23.976000 Hz
      3840x1600 px, 24.002001 Hz
      1920x1200 px, 59.993999 Hz
      1920x1080 px, 59.993999 Hz
      1600x1200 px, 59.993999 Hz
      1680x1050 px, 59.993999 Hz
      1280x1024 px, 59.993999 Hz
      1440x900 px, 59.993999 Hz
      1280x800 px, 59.993999 Hz
      1280x720 px, 59.993999 Hz
      1024x768 px, 60.004002 Hz
      800x600 px, 60.317001 Hz
      640x480 px, 60.000000 Hz
      640x480 px, 59.939999 Hz
    Position: 0,0
    Transform: normal
    Scale: 1.000000
    Adaptive Sync: enabled
  ```
  
</details>


`HZ: 143.998` (works)

<details>
  <summary>wlr-randr</summary>
  
  ```
  DP-3 "Dell Inc. Dell AW3821DW #HLAYMxgwABTR (DP-3)"
    Make: Dell Inc.
    Model: Dell AW3821DW
    Serial: #HLAYMxgwABTR
    Physical size: 880x370 mm
    Enabled: yes
    Modes:
      3840x1600 px, 59.993999 Hz (preferred)
      3840x1600 px, 143.998001 Hz (current)
      3840x1600 px, 120.044998 Hz
      3840x1600 px, 119.982002 Hz
      3840x1600 px, 99.970001 Hz
      3840x1600 px, 96.036003 Hz
      3840x1600 px, 84.971001 Hz
      3840x1600 px, 72.019997 Hz
      3840x1600 px, 60.007999 Hz
      3840x1600 px, 50.002998 Hz
      3840x1600 px, 48.008999 Hz
      3840x1600 px, 29.971001 Hz
      3840x1600 px, 30.000000 Hz
      3840x1600 px, 25.000999 Hz
      3840x1600 px, 23.976000 Hz
      3840x1600 px, 24.002001 Hz
      1920x1200 px, 59.993999 Hz
      1920x1080 px, 59.993999 Hz
      1600x1200 px, 59.993999 Hz
      1680x1050 px, 59.993999 Hz
      1280x1024 px, 59.993999 Hz
      1440x900 px, 59.993999 Hz
      1280x800 px, 59.993999 Hz
      1280x720 px, 59.993999 Hz
      1024x768 px, 60.004002 Hz
      800x600 px, 60.317001 Hz
      640x480 px, 60.000000 Hz
      640x480 px, 59.939999 Hz
    Position: 0,0
    Transform: normal
    Scale: 1.000000
    Adaptive Sync: enabled
  ```
  
</details>

`HZ: 143.99` (works)

<details>
  <summary>wlr-randr</summary>
  
  ```
  DP-3 "Dell Inc. Dell AW3821DW #HLAYMxgwABTR (DP-3)"
    Make: Dell Inc.
    Model: Dell AW3821DW
    Serial: #HLAYMxgwABTR
    Physical size: 880x370 mm
    Enabled: yes
    Modes:
      3840x1600 px, 59.993999 Hz (preferred)
      3840x1600 px, 143.998001 Hz (current)
      3840x1600 px, 120.044998 Hz
      3840x1600 px, 119.982002 Hz
      3840x1600 px, 99.970001 Hz
      3840x1600 px, 96.036003 Hz
      3840x1600 px, 84.971001 Hz
      3840x1600 px, 72.019997 Hz
      3840x1600 px, 60.007999 Hz
      3840x1600 px, 50.002998 Hz
      3840x1600 px, 48.008999 Hz
      3840x1600 px, 29.971001 Hz
      3840x1600 px, 30.000000 Hz
      3840x1600 px, 25.000999 Hz
      3840x1600 px, 23.976000 Hz
      3840x1600 px, 24.002001 Hz
      1920x1200 px, 59.993999 Hz
      1920x1080 px, 59.993999 Hz
      1600x1200 px, 59.993999 Hz
      1680x1050 px, 59.993999 Hz
      1280x1024 px, 59.993999 Hz
      1440x900 px, 59.993999 Hz
      1280x800 px, 59.993999 Hz
      1280x720 px, 59.993999 Hz
      1024x768 px, 60.004002 Hz
      800x600 px, 60.317001 Hz
      640x480 px, 60.000000 Hz
      640x480 px, 59.939999 Hz
    Position: 0,0
    Transform: normal
    Scale: 1.000000
    Adaptive Sync: enabled
  ```
  
</details>

`HZ: 143.9` (works)

<details>
  <summary>wlr-randr</summary>
  
  ```
DP-3 "Dell Inc. Dell AW3821DW #HLAYMxgwABTR (DP-3)"
  Make: Dell Inc.
  Model: Dell AW3821DW
  Serial: #HLAYMxgwABTR
  Physical size: 880x370 mm
  Enabled: yes
  Modes:
    3840x1600 px, 59.993999 Hz (preferred)
    3840x1600 px, 143.998001 Hz (current)
    3840x1600 px, 120.044998 Hz
    3840x1600 px, 119.982002 Hz
    3840x1600 px, 99.970001 Hz
    3840x1600 px, 96.036003 Hz
    3840x1600 px, 84.971001 Hz
    3840x1600 px, 72.019997 Hz
    3840x1600 px, 60.007999 Hz
    3840x1600 px, 50.002998 Hz
    3840x1600 px, 48.008999 Hz
    3840x1600 px, 29.971001 Hz
    3840x1600 px, 30.000000 Hz
    3840x1600 px, 25.000999 Hz
    3840x1600 px, 23.976000 Hz
    3840x1600 px, 24.002001 Hz
    1920x1200 px, 59.993999 Hz
    1920x1080 px, 59.993999 Hz
    1600x1200 px, 59.993999 Hz
    1680x1050 px, 59.993999 Hz
    1280x1024 px, 59.993999 Hz
    1440x900 px, 59.993999 Hz
    1280x800 px, 59.993999 Hz
    1280x720 px, 59.993999 Hz
    1024x768 px, 60.004002 Hz
    800x600 px, 60.317001 Hz
    640x480 px, 60.000000 Hz
    640x480 px, 59.939999 Hz
  Position: 0,0
  Transform: normal
  Scale: 1.000000
  Adaptive Sync: enable
  ```
  
</details>

`HZ: 143` (not valid as expected)
  
<details>
  <summary>wlr-randr</summary>
  
  ```
  DP-3 "Dell Inc. Dell AW3821DW #HLAYMxgwABTR (DP-3)"
    Make: Dell Inc.
    Model: Dell AW3821DW
    Serial: #HLAYMxgwABTR
    Physical size: 880x370 mm
    Enabled: yes
    Modes:
      3840x1600 px, 59.993999 Hz (preferred, current)
      3840x1600 px, 143.998001 Hz
      3840x1600 px, 120.044998 Hz
      3840x1600 px, 119.982002 Hz
      3840x1600 px, 99.970001 Hz
      3840x1600 px, 96.036003 Hz
      3840x1600 px, 84.971001 Hz
      3840x1600 px, 72.019997 Hz
      3840x1600 px, 60.007999 Hz
      3840x1600 px, 50.002998 Hz
      3840x1600 px, 48.008999 Hz
      3840x1600 px, 29.971001 Hz
      3840x1600 px, 30.000000 Hz
      3840x1600 px, 25.000999 Hz
      3840x1600 px, 23.976000 Hz
      3840x1600 px, 24.002001 Hz
      1920x1200 px, 59.993999 Hz
      1920x1080 px, 59.993999 Hz
      1600x1200 px, 59.993999 Hz
      1680x1050 px, 59.993999 Hz
      1280x1024 px, 59.993999 Hz
      1440x900 px, 59.993999 Hz
      1280x800 px, 59.993999 Hz
      1280x720 px, 59.993999 Hz
      1024x768 px, 60.004002 Hz
      800x600 px, 60.317001 Hz
      640x480 px, 60.000000 Hz
      640x480 px, 59.939999 Hz
    Position: 0,0
    Transform: normal
    Scale: 1.000000
    Adaptive Sync: enabled
  ```
  
</details>

---

`HZ: 120.044998` (works)

<details>
  <summary>wlr-randr</summary>

  ```
  DP-3 "Dell Inc. Dell AW3821DW #HLAYMxgwABTR (DP-3)"
    Make: Dell Inc.
    Model: Dell AW3821DW
    Serial: #HLAYMxgwABTR
    Physical size: 880x370 mm
    Enabled: yes
    Modes:
      3840x1600 px, 59.993999 Hz (preferred)
      3840x1600 px, 143.998001 Hz
      3840x1600 px, 120.044998 Hz (current)
      3840x1600 px, 119.982002 Hz
      3840x1600 px, 99.970001 Hz
      3840x1600 px, 96.036003 Hz
      3840x1600 px, 84.971001 Hz
      3840x1600 px, 72.019997 Hz
      3840x1600 px, 60.007999 Hz
      3840x1600 px, 50.002998 Hz
      3840x1600 px, 48.008999 Hz
      3840x1600 px, 29.971001 Hz
      3840x1600 px, 30.000000 Hz
      3840x1600 px, 25.000999 Hz
      3840x1600 px, 23.976000 Hz
      3840x1600 px, 24.002001 Hz
      1920x1200 px, 59.993999 Hz
      1920x1080 px, 59.993999 Hz
      1600x1200 px, 59.993999 Hz
      1680x1050 px, 59.993999 Hz
      1280x1024 px, 59.993999 Hz
      1440x900 px, 59.993999 Hz
      1280x800 px, 59.993999 Hz
      1280x720 px, 59.993999 Hz
      1024x768 px, 60.004002 Hz
      800x600 px, 60.317001 Hz
      640x480 px, 60.000000 Hz
      640x480 px, 59.939999 Hz
    Position: 0,0
    Transform: normal
    Scale: 1.000000
    Adaptive Sync: enabled
  ```
  
</details>

`HZ: 120` (works)

<details>
  <summary>wlr-randr</summary>
  
  ```
  DP-3 "Dell Inc. Dell AW3821DW #HLAYMxgwABTR (DP-3)"
    Make: Dell Inc.
    Model: Dell AW3821DW
    Serial: #HLAYMxgwABTR
    Physical size: 880x370 mm
    Enabled: yes
    Modes:
      3840x1600 px, 59.993999 Hz (preferred)
      3840x1600 px, 143.998001 Hz
      3840x1600 px, 120.044998 Hz (current)
      3840x1600 px, 119.982002 Hz
      3840x1600 px, 99.970001 Hz
      3840x1600 px, 96.036003 Hz
      3840x1600 px, 84.971001 Hz
      3840x1600 px, 72.019997 Hz
      3840x1600 px, 60.007999 Hz
      3840x1600 px, 50.002998 Hz
      3840x1600 px, 48.008999 Hz
      3840x1600 px, 29.971001 Hz
      3840x1600 px, 30.000000 Hz
      3840x1600 px, 25.000999 Hz
      3840x1600 px, 23.976000 Hz
      3840x1600 px, 24.002001 Hz
      1920x1200 px, 59.993999 Hz
      1920x1080 px, 59.993999 Hz
      1600x1200 px, 59.993999 Hz
      1680x1050 px, 59.993999 Hz
      1280x1024 px, 59.993999 Hz
      1440x900 px, 59.993999 Hz
      1280x800 px, 59.993999 Hz
      1280x720 px, 59.993999 Hz
      1024x768 px, 60.004002 Hz
      800x600 px, 60.317001 Hz
      640x480 px, 60.000000 Hz
      640x480 px, 59.939999 Hz
    Position: 0,0
    Transform: normal
    Scale: 1.000000
    Adaptive Sync: enabled
  ```
  
</details>

`HZ: 119.982002` (works)

<details>
  <summary>wlr-randr</summary>

  ```
  DP-3 "Dell Inc. Dell AW3821DW #HLAYMxgwABTR (DP-3)"
    Make: Dell Inc.
    Model: Dell AW3821DW
    Serial: #HLAYMxgwABTR
    Physical size: 880x370 mm
    Enabled: yes
    Modes:
      3840x1600 px, 59.993999 Hz (preferred)
      3840x1600 px, 143.998001 Hz
      3840x1600 px, 120.044998 Hz
      3840x1600 px, 119.982002 Hz (current)
      3840x1600 px, 99.970001 Hz
      3840x1600 px, 96.036003 Hz
      3840x1600 px, 84.971001 Hz
      3840x1600 px, 72.019997 Hz
      3840x1600 px, 60.007999 Hz
      3840x1600 px, 50.002998 Hz
      3840x1600 px, 48.008999 Hz
      3840x1600 px, 29.971001 Hz
      3840x1600 px, 30.000000 Hz
      3840x1600 px, 25.000999 Hz
      3840x1600 px, 23.976000 Hz
      3840x1600 px, 24.002001 Hz
      1920x1200 px, 59.993999 Hz
      1920x1080 px, 59.993999 Hz
      1600x1200 px, 59.993999 Hz
      1680x1050 px, 59.993999 Hz
      1280x1024 px, 59.993999 Hz
      1440x900 px, 59.993999 Hz
      1280x800 px, 59.993999 Hz
      1280x720 px, 59.993999 Hz
      1024x768 px, 60.004002 Hz
      800x600 px, 60.317001 Hz
      640x480 px, 60.000000 Hz
      640x480 px, 59.939999 Hz
    Position: 0,0
    Transform: normal
    Scale: 1.000000
    Adaptive Sync: enabled
  ```
  
</details>

HZ: 119 (not valid as expected)

<details>
  <summary>wlr-randr</summary>

  ```
  DP-3 "Dell Inc. Dell AW3821DW #HLAYMxgwABTR (DP-3)"
    Make: Dell Inc.
    Model: Dell AW3821DW
    Serial: #HLAYMxgwABTR
    Physical size: 880x370 mm
    Enabled: yes
    Modes:
      3840x1600 px, 59.993999 Hz (preferred, current)
      3840x1600 px, 143.998001 Hz
      3840x1600 px, 120.044998 Hz
      3840x1600 px, 119.982002 Hz
      3840x1600 px, 99.970001 Hz
      3840x1600 px, 96.036003 Hz
      3840x1600 px, 84.971001 Hz
      3840x1600 px, 72.019997 Hz
      3840x1600 px, 60.007999 Hz
      3840x1600 px, 50.002998 Hz
      3840x1600 px, 48.008999 Hz
      3840x1600 px, 29.971001 Hz
      3840x1600 px, 30.000000 Hz
      3840x1600 px, 25.000999 Hz
      3840x1600 px, 23.976000 Hz
      3840x1600 px, 24.002001 Hz
      1920x1200 px, 59.993999 Hz
      1920x1080 px, 59.993999 Hz
      1600x1200 px, 59.993999 Hz
      1680x1050 px, 59.993999 Hz
      1280x1024 px, 59.993999 Hz
      1440x900 px, 59.993999 Hz
      1280x800 px, 59.993999 Hz
      1280x720 px, 59.993999 Hz
      1024x768 px, 60.004002 Hz
      800x600 px, 60.317001 Hz
      640x480 px, 60.000000 Hz
      640x480 px, 59.939999 Hz
    Position: 0,0
    Transform: normal
    Scale: 1.000000
    Adaptive Sync: enabled
  ```
  
</details>

**Comment by alex-courtis** [comment]
Thank you for the detailed testing!

I'll look into a release for this shortly.

**Comment by GrabbenD** [comment]
Sway has just also implemented this feature 🙂 
https://github.com/swaywm/sway/pull/7783 (ref: https://github.com/swaywm/sway/issues/7599#issuecomment-1765212668)

**Comment by alex-courtis** [comment]
> Sway has just also implemented this feature 🙂 [swaywm/sway#7783](https://github.com/swaywm/sway/pull/7783) (ref: [swaywm/sway#7599 (comment)](https://github.com/swaywm/sway/issues/7599#issuecomment-1765212668))

Nice!
